sample/sample-with-di/helloworldfeature/src/main/java/com/dropbox/kaiken/sample/advancedsample/helloworldfeature/ComposeBindings.kt

```diff
@@ -0,0 +1,204 @@
+    arguments: List<NamedNavArgument> = emptyList(),
+    deepLinks: List<NavDeepLink> = emptyList(),
+    crossinline content: @Composable AuthRequiredScreenComponent.(NavBackStackEntry, T) -> Unit,
+) {
+
+    val kaikenAwareContent: @Composable (NavBackStackEntry) -> Unit = { entry ->
+        val injector: AuthRequiredScreenComponent.ScreenParentComponent =
+            LocalDResolver.current.resolveDependencyProvider()
+
+        val retained = entry.retain { injector.authRequiredScreenComponent() }
+        retainComponentAndSetContent(retained, entry, content)
+    }
+    composable(route, arguments, deepLinks, kaikenAwareContent)
+}
+
+inline fun <reified T : BasePresenter> NavGraphBuilder.authOptionalComposable(
+    route: String,
+    arguments: List<NamedNavArgument> = emptyList(),
+    deepLinks: List<NavDeepLink> = emptyList(),
+    crossinline content: @Composable AuthOptionalScreenComponent.(NavBackStackEntry, T) -> Unit,
+) {
+    val kaikenAwareContent: @Composable (NavBackStackEntry) -> Unit = { entry ->
+        val injector: AuthOptionalScreenComponent.ScreenParentComponent =
+            LocalDResolver.current.resolveDependencyProvider()
+
+        val retained = entry.retain { injector.authOptionalScreenComponent() }
+        retainComponentAndSetContent(retained, entry, content)
+
+    }
+    composable(route, arguments, deepLinks, kaikenAwareContent)
+}
+
+inline fun <reified T : BasePresenter> NavGraphBuilder.authAwareComposable(
+    route: String,
+    arguments: List<NamedNavArgument> = emptyList(),
+    deepLinks: List<NavDeepLink> = emptyList(),
+    crossinline content: @Composable Injector.(NavBackStackEntry, T) -> Unit,
+) {
+    val kaikenAwareContent: @Composable (NavBackStackEntry) -> Unit = { entry ->
+
+        if (LocalContext.current is AuthRequiredActivity) {
+            val injector: AuthRequiredScreenComponent.ScreenParentComponent =
+                LocalDResolver.current.resolveDependencyProvider()
+            val retained = entry.retain { injector.authRequiredScreenComponent() }
+            retainComponentAndSetContent(retained, entry, content)
+        } else if (LocalContext.current is AuthOptionalActivity) {
+            val injector: AuthOptionalScreenComponent.ScreenParentComponent =
+                LocalDResolver.current.resolveDependencyProvider()
+            val retained = entry.retain { injector.authOptionalScreenComponent() }
+            retainComponentAndSetContent(retained, entry, content)
+        }
+    }
+    composable(route, arguments, deepLinks, kaikenAwareContent)
+}
+
+@Composable
+inline fun <reified T : BasePresenter, reified V:Injector> retainComponentAndSetContent(
+    retained: V,
+    entry: NavBackStackEntry,
+    crossinline content: @Composable() (V.(NavBackStackEntry, T) -> Unit)
+) {
+    CompositionLocalProvider(LocalComponent provides retained) {
+        val presenter: T =
+            retained.cast<PresenterProvider>().presenters().filterIsInstance<T>()
```

Comment:
It feels like this could be a perf bottleneck if there are lots of presenters (which I'd expect there to be).  Can we use a `@ClassKey` on the presenter multibinding to bind them into a map keyed on the type?

Reply:
agreed, we can think about rewriting as we scale to instead by a map multibinding